src/settings.ts

```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+import * as VSCode from 'vscode';
+import { ExtensionContext } from 'vscode';
+
+const SONARLINT_CATEGORY = 'sonarlint';
+const CONNECTIONS_SECTION = 'connectedMode.connections';
+const SONARQUBE = 'sonarqube';
+const SONARQUBE_CONNECTIONS_CATEGORY = `${SONARLINT_CATEGORY}.${CONNECTIONS_SECTION}.${SONARQUBE}`;
+
+async function hasUnmigratedConnections(sqConnections: SqConnection[],
+                                        settingsService: ConnectionSettingsService): Promise<boolean> {
+  for (const connection of sqConnections) {
+    if (!await settingsService.hasTokenForServer(connection.serverUrl)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+export async function migrateConnectedModeSettings(settings: VSCode.WorkspaceConfiguration, settingsService: ConnectionSettingsService) {
+  const sqConnections = settings.get<SqConnection[]>(`${CONNECTIONS_SECTION}.${SONARQUBE}`);
+  if (await hasUnmigratedConnections(sqConnections, settingsService)) {
+    suggestMigrationToSecureStorage(sqConnections, settingsService);
+  }
+}
+
+async function suggestMigrationToSecureStorage(sqConnections: SqConnection[], settingsService: ConnectionSettingsService) {
+  const remindMeLaterAction = 'Ask me later';
+  const migrateToSecureStorageAction = 'Migrate connection settings to secure storage';
+  const message = `SonarLint found SonarQube token in settings file. Do you want to migrate them to secure storage?`;
+  const selection = await VSCode.window.showWarningMessage(message, migrateToSecureStorageAction, remindMeLaterAction);
+  switch (selection) {
+    case migrateToSecureStorageAction:
+      await settingsService.addTokensFromSettingsToSecureStorage(sqConnections);
+      break;
+    case remindMeLaterAction:
+  }
+}
+
+export function getSonarLintConfiguration(): VSCode.WorkspaceConfiguration {
+  return VSCode.workspace.getConfiguration(SONARLINT_CATEGORY);
+}
+
+
+export class ConnectionSettingsService {
+
+  private static instance: ConnectionSettingsService;
+
+  constructor(private readonly secretStorage: VSCode.SecretStorage) {
+  }
+
+  static init(context: ExtensionContext): void {
+    ConnectionSettingsService.instance = new ConnectionSettingsService(context.secrets);
+  }
+
+  static get getInstance(): ConnectionSettingsService {
+    return ConnectionSettingsService.instance;
+  }
+
+  /**
+   *
+   * @param serverId SonarQube URL or SonarCloud organization ID
+   * @param token auth token
+   */
+  async storeServerToken(serverId: string, token: string): Promise<void> {
+    if (token) {
+      this.secretStorage.store(serverId, token);
+    }
+  }
+
+  async getServerToken(serverId: string): Promise<string | undefined> {
+    return this.secretStorage.get(serverId);
+  }
+
+  async hasTokenForServer(serverId: string): Promise<boolean> {
+    return (await this.getServerToken(serverId)) !== undefined;
+  }
+
+  async deleteTokenForServer(serverId: string): Promise<void> {
```

Comment:
What is `serverId` ? Does it refer to `connectionId` or `serverUrl` from current context?

Reply:
Yes.